src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepDynamicContextRJRTest.java

```diff
@@ -0,0 +1,127 @@
+            iar.createAgent(rjr, InboundAgentRule.Options.newBuilder().name("J").label("mib").color(PrefixedOutputStream.Color.YELLOW).webSocket().build());
+            rjr.runRemotely(ExecutorStepDynamicContextRJRTest::setupJobAndStart);
+            rjr.stopJenkinsForcibly();
+            rjr.startJenkins();
+            rjr.runRemotely(ExecutorStepDynamicContextRJRTest::assertQueueItems);
+        }
+    }
+
+    private static void assertQueueItems(JenkinsRule r) throws Throwable {
+        var p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+        var b = p.getBuildByNumber(1);
+        var items = await("Waiting for agent J to reconnect").atMost(Duration.ofSeconds(30)).until(() -> r.jenkins.getQueue().getItems(), arrayWithSize(1));
+        LOGGER.info("Items in queue: " + Arrays.asList(items));
+        var actions = await().until(() -> b.getActions(InputAction.class), hasSize(1));
+        proceed(actions, "Complete branch 1 ?", p.getName() + "#" + b.number);
+        actions = await().until(() -> b.getActions(InputAction.class), allOf(iterableWithSize(1), hasItem(new InputActionIsWaitingForInput())));
+        r.waitForMessage("Complete branch 2 ?", b);
+        proceed(actions, "Complete branch 2 ?", p.getName() + "#" + b.number);
+        r.waitForCompletion(b);
+    }
+
+    private static class InputActionIsWaitingForInput extends TypeSafeMatcher<InputAction> {
+
+        @Override
+        protected boolean matchesSafely(InputAction inputAction) {
+            try {
+                return inputAction.isWaitingForInput();
+            } catch (Exception e) {
+                return false;
+            }
+        }
+
+        @Override
+        public void describeTo(Description description) {
+            description.appendText("is waiting for input");
+        }
+    }
+
+    private static void proceed(List<InputAction> actions, String inputMessage, String name) throws InterruptedException, TimeoutException, IOException {
+        for (var action : actions) {
+            if (action.getExecutions().isEmpty()) {
+                continue;
+            }
+            var inputStepExecution = action.getExecutions().get(0);
+            if (inputMessage.equals(inputStepExecution.getDisplayName())) {
+                LOGGER.info(() -> "proceeding " + name);
+                inputStepExecution.proceed(null);
+                break;
+            }
+        }
+    }
+
+    private static void setupJobAndStart(JenkinsRule r) throws Exception {
+        var p = r.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("parallel 'Branch 1': { node('mib') { input 'Complete branch 1 ?' } }, 'Branch 2': { sleep 1; node('mib') { input 'Complete branch 2 ?' } }", true));
+        var b = p.scheduleBuild2(0).waitForStart();
+        r.waitForMessage("Complete branch 1 ?", b);
+        assertTrue(b.isBuilding());
+        await().until(() -> r.jenkins.getQueue().getItems(), arrayWithSize(1));
```

Comment:
Looks like a race condition—will the test still run OK if branch 2 gets as far as starting `node` by this point?